Erc721 votes and general Votes component (#1114)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/api/governance.adoc

Emitted when the minimum delay for future operations is modified.

== Votes

> The `VotesComponent` provides a flexible system for tracking voting power and delegation. This system allows users to delegate their voting power to other addresses, enabling more active participation in governance.

Reviewer: ```suggestion
The `VotesComponent` provides a flexible system for tracking voting power. This system allows users to delegate their voting power to other addresses, enabling more active participation in governance.
```

Author: why not mention delegation? Seems relevant

Reviewer: It is mentioned in the next sentence, while in this one it sounds like it is also a "flexible system for tracking delegation" (the and), while the main purpose is to track voting power, while delegation is a feature.

.Embeddable Implementations
--
[.sub-index#VotesComponent-Embeddable-Impls-VotesImpl]
.VotesImpl
* xref:#VotesComponent-get_votes[`++get_votes(self, account)++`]
* xref:#VotesComponent-get_past_votes[`++get_past_votes(self, account, timepoint)++`]
* xref:#VotesComponent-get_past_total_supply[`++get_past_total_supply(self, timepoint)++`]
* xref:#VotesComponent-delegates[`++delegates(self, account)++`]
* xref:#VotesComponent-delegate[`++delegate(self, delegatee)++`]
* xref:#VotesComponent-delegate_by_sig[`++delegate_by_sig(self, delegator, delegatee, nonce, expiry, signature)++`]
--

[.contract-index]
.Internal implementations
--
.InternalImpl
* xref:#VotesComponent-_delegate[`++_delegate(self, account, delegatee)++`]
* xref:#VotesComponent-move_delegate_votes[`++move_delegate_votes(self, from, to, amount)++`]
* xref:#VotesComponent-transfer_voting_units[`++transfer_voting_units(self, from, to, amount)++`]
--

[.contract-index]
.Events
--
* xref:#VotesComponent-DelegateChanged[`++DelegateChanged(delegator, from_delegate, to_delegate)++`]
* xref:#VotesComponent-DelegateVotesChanged[`++DelegateVotesChanged(delegate, previous_votes, new_votes)++`]
--

> [#VotesComponent-VotingUnitsTrait-Implementations]
> ==== ERC20VotesImpl
> 
> [.contract-item]
> [[VotesComponent-ERC20VotesImpl-get_voting_units]]

Reviewer 2: ```suggestion
[#VotesComponent-ERC20VotesImpl]
==== ERC20VotesImpl

[.contract-item]
[[VotesComponent-ERC20VotesImpl-get_voting_units]]
```
There's no section for VotingUnitTrait-Implementations so I think we can remove that in favor of ERC20VotesImpl. This allows the `get_voting_units` ID to logically follow

==== `[.contract-item-name]#++get_voting_units++#++(self: @ComponentState<TContractState>, account: ContractAddress) → u256++` [.item-kind]#internal#

Returns the number of voting units for a given account.

This implementation is specific to ERC20 tokens, where the balance
of tokens directly represents the number of voting units.

NOTE: This implementation will work out of the box if the ERC20 component
is implemented in the final contract.

WARNING: This implementation assumes tokens map to voting units 1:1.
Any deviation from this formula when transferring voting units (e.g. by using hooks)
may compromise the internal vote accounting.

> ==== ERC721VotesImpl

Reviewer 2: We should add an ID here that follows the ERC20VotesImpl approach. Even if we go with something different than the above suggestion, this should still follow w/e pattern we use

Returns the delegate that `account` has chosen.

[.contract-item]
[[VotesComponent-delegate]]
==== `[.contract-item-name]#++delegate++#++(ref self: ComponentState<TContractState>, delegatee: ContractAddress)++` [.item-kind]#external#

Delegates votes from the sender to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-delegate_by_sig]]
==== `[.contract-item-name]#++delegate_by_sig++#++(ref self: ComponentState<TContractState>, delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Span<felt252>)++` [.item-kind]#external#

Delegates votes from `delegator` to `delegatee` through a SNIP12 message signature validation.

Requirements:

- `expiry` must not be in the past.
- `nonce` must match the account's current nonce.
- `delegator` must implement `SRC6::is_valid_signature`.
- `signature` should be valid for the message hash.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[#VotesComponent-Internal-functions]
==== Internal functions

> 

Author: Nice catch! Looking forward to when we have a place to publish the inline docs and don't have to keep these 2 in sync.

[.contract-item]
[[VotesComponent-_delegate]]
==== `[.contract-item-name]#++_delegate++#++(ref self: ComponentState<TContractState>, account: ContractAddress, delegatee: ContractAddress)++` [.item-kind]#internal#

Delegates all of ``account``'s voting units to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-move_delegate_votes]]
==== `[.contract-item-name]#++move_delegate_votes++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Moves delegated votes from one delegate to another.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-transfer_voting_units]]
==== `[.contract-item-name]#++transfer_voting_units++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Transfers, mints, or burns voting units.

To register a mint, `from` should be zero. To register a burn, `to`
should be zero. Total supply of voting units will be adjusted with mints and burns.

> CAUTION: If voting units are a function of an underlying transferrable asset (like a token), this function should be

Author: What about this?
```
CAUTION: If voting units are based on an underlying transferable asset (like a token), you must call this function every time the asset is transferred to keep the internal voting power accounting in sync. For ERC20 and ERC721 tokens, this is typically handled using hooks.
```

I'm changing it here and in the underlying component docs just so we can resolve sooner, but feel free to suggest something else and we can adapt it.

> CAUTION: If voting units are a function of an underlying transferrable asset (like a token), this function should be

Reviewer 2: Nice! Yeah, sounds good to me

Delegates votes from the sender to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-delegate_by_sig]]
==== `[.contract-item-name]#++delegate_by_sig++#++(ref self: ComponentState<TContractState>, delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Span<felt252>)++` [.item-kind]#external#

Delegates votes from `delegator` to `delegatee` through a SNIP12 message signature validation.

Requirements:

- `expiry` must not be in the past.
- `nonce` must match the account's current nonce.
- `delegator` must implement `SRC6::is_valid_signature`.
- `signature` should be valid for the message hash.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[#VotesComponent-Internal-functions]
==== Internal functions

[.contract-item]
[[VotesComponent-get_total_supply]]
==== `[.contract-item-name]#++get_total_supply++#++(self: @ComponentState<TContractState>) → u256++` [.item-kind]#internal#

Returns the current total supply of votes.

[.contract-item]
[[VotesComponent-_delegate]]

> ==== `[.contract-item-name]#++_delegate++#++(ref self: ComponentState<TContractState>, account: ContractAddress, delegatee: ContractAddress)++` [.item-kind]#internal#

Reviewer 2: Sorry, I should have mentioned this when I brought up adding the internal fns. I think we should move the underscored methods to the end which is what we're doing with the other APIs. For consistency, I'd change the fn order in the code as well so they match

[.contract-item]
[[VotesComponent-get_past_total_supply]]
==== `[.contract-item-name]#++get_past_total_supply++#++(self: @ComponentState<TContractState>, timepoint: u64) → u256++` [.item-kind]#external#

Returns the total supply of votes available at a specific moment in the past.

NOTE: This value is the sum of all available votes, which is not necessarily the sum of all delegated votes.
Votes that have not been delegated are still part of total supply, even though they would not participate in a
vote.

Requirements:

- `timepoint` must be in the past.

[.contract-item]
[[VotesComponent-delegates]]
==== `[.contract-item-name]#++delegates++#++(self: @ComponentState<TContractState>, account: ContractAddress) → ContractAddress++` [.item-kind]#external#

Returns the delegate that `account` has chosen.

[.contract-item]
[[VotesComponent-delegate]]
==== `[.contract-item-name]#++delegate++#++(ref self: ComponentState<TContractState>, delegatee: ContractAddress)++` [.item-kind]#external#

Delegates votes from the sender to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-delegate_by_sig]]

> ==== `[.contract-item-name]#++delegate_by_sig++#++(ref self: ComponentState<TContractState>, delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Array<felt252>)++` [.item-kind]#external#

Reviewer: +1 to using Span instead.

> ==== `[.contract-item-name]#++delegate_by_sig++#++(ref self: ComponentState<TContractState>, delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Array<felt252>)++` [.item-kind]#external#

Author: When looking into this I realized we would need to change the type that `ISRC6::is_valid_signature` accepts(or convert from span to array - which doesn't make a lot of sense), and since it is part of the [SNIP-6](https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-6.md) standard does not seem like the best idea either.
Thoughts on this? Should we leave it with Array?

> ==== `[.contract-item-name]#++delegate_by_sig++#++(ref self: ComponentState<TContractState>, delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Array<felt252>)++` [.item-kind]#external#

Reviewer: Similar to what is implemented in SRC9, we should accept Span and just call `.into()` for is_valid_signature, and this should be less expensive than receiving `Array<felt252>` instead, so we don't have a reason to compromise more interfaces when we should be prioritizing `Span<felt252>` over `Array<felt252>` in external function inputs (because of the deserialization required for dispatchers).

==== Internal functions

[.contract-item]
[[VotesComponent-_delegate]]
==== `[.contract-item-name]#++_delegate++#++(ref self: ComponentState<TContractState>, account: ContractAddress, delegatee: ContractAddress)++` [.item-kind]#internal#

Delegates all of ``account``'s voting units to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-move_delegate_votes]]
==== `[.contract-item-name]#++move_delegate_votes++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Moves delegated votes from one delegate to another.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-transfer_voting_units]]
==== `[.contract-item-name]#++transfer_voting_units++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Transfers, mints, or burns voting units.

To register a mint, `from` should be zero. To register a burn, `to`
should be zero. Total supply of voting units will be adjusted with mints and burns.

> May emit one or two {DelegateVotesChanged} events.

Reviewer 2: Perhaps, it's worth adding another NOTE for this in the beginning of the component section as well?

> May emit one or two {DelegateVotesChanged} events.

Author: Good call, this is the trickiest part of the component IMO. I added it here and as IMPORTANT to the component section in the main `governance.adoc`

==== `[.contract-item-name]#++delegate_by_sig++#++(delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Array<felt252>)++` [.item-kind]#external#

Delegates votes from `delegator` to `delegatee`.

> 

Reviewer: ```suggestion
```

[.contract]
[[VotesComponent]]
=== `++VotesComponent++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.17.0/packages/governance/src/votes/votes.cairo[{github-icon},role=heading-link]

[.hljs-theme-dark]
```cairo
use openzeppelin_governance::votes::VotesComponent;
```

> By default, token balance does not account for voting power. This makes transfers cheaper. The downside is that it requires users to delegate to themselves in order to activate checkpoints and have their voting power tracked.
> 
> NOTE: When using this module, your contract must implement the {VotingUnitsTrait}. For convinience, this is done automatically for `ERC20` and `ERC721` tokens.

Reviewer: ```suggestion
The Votes component provides a flexible system for tracking and delegating voting power.
It is currently implemented for ERC20 and ERC721 tokens. An account can delegate
their voting power to a representative, that will pool delegated voting units from different
delegators and can then use it to vote in decisions. Voting power must be delegated to be
counted, and an account must delegate to itself if it wishes to vote directly without a trusted
representative.

When integrating the Votes component, the {VotingUnitsTrait} must be implemented to get the
voting units for a given account as a function of the implementing contract. For simplicity,
this module already provides two implementations for ERC20 and ERC721 tokens, which will work
out of the box if the respective components are integrated.

NOTE: ERC20 and ERC721 tokens implementing this component must call {transfer_voting_units}
whenever a transfer, mint, or burn operation is performed. Hooks can be leveraged for this
purpose.
```

Reviewer: Why not include the full component description here? Also, splitting into multiple lines as suggested before.

Author: All of this is included in the main `Votes` page in `governance.adoc`. Do you suggest I copy/paste what is in there? It feels like unnecessary repetition. 
My expectation is that someone reading the docs looks at the main `Votes` section and then comes to the api for details on function signature, etc. It was ok to have more info in the api section(and I don't think we had much more TBH) when `ERC20Votes` was just an extension.

Reviewer: @andrew-fleming , marto and me had this conversation before, and this repetition may be helpful since users in practice don't have to go to the other page first, and documentation should be as easy to understand and follow as possible. Users may come directly to the API page from the READMEs shown in the registry for example, and we shouldn't expect them to go then to the main page and later return here. I vote we add this here too.

Author: I understand this... but it also defeats the purpose of having two separate pages and incurs too much repetition. If we were to include things from the main page into api I would say we should only have a single page then.
I was also checking other examples(e.g. finance, [timelock](https://docs.openzeppelin.com/contracts-cairo/0.17.0/api/governance) and even [ERC20](https://docs.openzeppelin.com/contracts-cairo/0.17.0/api/erc20)) and we haven't been following this patterns, so this doesn't seem very consistent. One option that could work is what we did for ERC20 and ad a NOTE with a reference to the main docs.
WDYT?

Reviewer: You are right, we don't need a full description here matching what we explained on the main page. Maybe let's rephrase the sentence a bit though, since starting with "By default, token balance does not account for voting power" feels like some context is skipped. A short description like in the other modules, and a couple of notes should suffice.

Author: Awesome, added a short description at the beginning of the paragraph

[.contract-index#VotesComponent-Embeddable-Impls]
.Embeddable Implementations
--
[.sub-index#VotesComponent-Embeddable-Impls-VotesImpl]
.VotesImpl
* xref:#VotesComponent-get_votes[`++get_votes(self, account)++`]
* xref:#VotesComponent-get_past_votes[`++get_past_votes(self, account, timepoint)++`]
* xref:#VotesComponent-get_past_total_supply[`++get_past_total_supply(self, timepoint)++`]
* xref:#VotesComponent-delegates[`++delegates(self, account)++`]
* xref:#VotesComponent-delegate[`++delegate(self, delegatee)++`]
* xref:#VotesComponent-delegate_by_sig[`++delegate_by_sig(self, delegator, delegatee, nonce, expiry, signature)++`]
--

[.contract-index]

> .Internal implementations

Reviewer: We should add a `VotingUnitsTrait implementations` section so users are aware of the implementations provided by default with the component, and also can see the comments explaining the formula followed for internal accounting.

Author: I added it and I'm not sure I love the result. Can you take a look and let me know if this is what you had in mind? Happy to iterate over this if needed.

Reviewer: Some notes:

- The `.VotingUnitsTrait implementations` should be the first section before `Embeddable implementations` for consistency with other pages having custom extra items.
The NOTE splitting the sections should not appear here but in the section expanding the function descriptions below (which we still need to add).
- We need to add the section below with the expanded descriptions (same comment that the function has in the code) and should add that they will work out-of-box if the corresponding component is implemented.

Reviewer: +1. I vote for `VOTING UNITS TRAIT IMPLEMENTATIONS`.

Author: Sorry, really struggled to understand what you meant here. So I followed an example from another doc(vesting wallet in this case) and went ahead with that pattern, where we have the trait implementation section before the Embeddable Implementations and we just link to the trait section. I think this makes it clear, since that section includes a more detailed description and even a WARNING, but feel free to edit and commit if you feel something is missing.

Reviewer: VestingWallet is a preset which is not documented in the same way. The updates you did are great, the only thing missing is a section in the component with the functions explained (copy pasted the impl function comment), and the header section linking to that instead.

Reviewer: Currently the first section below the header is `[#VotesComponent-Functions]`, we need to add a [#VotesComponent-VotingUnitsTrait-Implementations] above that, with the two impls names as sections, each of them with the function as sub-item and the comment as descriptor.

==== Internal functions

[.contract-item]
[[VotesComponent-_delegate]]
==== `[.contract-item-name]#++_delegate++#++(ref self: ComponentState<TContractState>, account: ContractAddress, delegatee: ContractAddress)++` [.item-kind]#internal#

Delegates all of ``account``'s voting units to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-move_delegate_votes]]
==== `[.contract-item-name]#++move_delegate_votes++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Moves delegated votes from one delegate to another.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-transfer_voting_units]]
==== `[.contract-item-name]#++transfer_voting_units++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Transfers, mints, or burns voting units.

To register a mint, `from` should be zero. To register a burn, `to`
should be zero. Total supply of voting units will be adjusted with mints and burns.

> Special care must be taken to ensure that this function is called correctly. In `ERC20` and `ERC721` this is done usually via the hooks.

Reviewer: ```suggestion
Special care must be taken to ensure that this function is called correctly. In `ERC20` and `ERC721` components this is done usually via the hooks.
```

Author: Well... for people implementing this won't be a component, it will be a contract in any case. What do you think about changing components for contracts?

May emit one or two {DelegateVotesChanged} events.

[#VotesComponent-Events]
==== Events

[.contract-item]
[[VotesComponent-DelegateChanged]]
==== `[.contract-item-name]#++DelegateChanged++#++(delegator: ContractAddress, from_delegate: ContractAddress, to_delegate: ContractAddress)++` [.item-kind]#event#

Emitted when an account changes their delegate.

[.contract-item]
[[VotesComponent-DelegateVotesChanged]]
==== `[.contract-item-name]#++DelegateVotesChanged++#++(delegate: ContractAddress, previous_votes: u256, new_votes: u256)++` [.item-kind]#event#

Emitted when a token transfer or delegate change results in changes to a delegate's number of votes.

[.contract]
[[VotingUnitsTrait]]
=== `++VotingUnitsTrait++` link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.17.0/packages/governance/src/votes/votes.cairo[{github-icon},role=heading-link]

```cairo
pub trait VotingUnitsTrait<TState> {
    fn get_voting_units(self: @TState, account: ContractAddress) -> u256;
}
```

> This trait must be implemented for tokens that want to use the VotesComponent. It is already implemented for ERC20 and ERC721.

Reviewer: ```suggestion
A trait that must be implemented when integrating {VotesComponent} into a contract. It offers a mechanism to retrieve the number of voting units for a given account at the current time.
```

[.contract-index]
.Functions
--
* xref:#VotingUnitsTrait-get_voting_units[`++get_voting_units(self, account)++`]
--

> [#TokenVotesTrait-Functions]

Reviewer: ```suggestion
[#VotingUnitsTrait-Functions]
```

Author: missed this one!

==== Functions

[.contract-item]
[[VotingUnitsTrait-get_voting_units]]
==== `[.contract-item-name]#++get_voting_units++#++(self: @TState, account: ContractAddress) → u256++` [.item-kind]#external#

> Returns the number of voting units for a given account. For ERC20, this is typically the token balance. For ERC721, this is typically the number of tokens owned.

Reviewer: ```suggestion
Returns the number of voting units for a given account. For ERC20, this is typically the token balance. For ERC721, this is typically the number of tokens owned.

WARNING: While any formula can be used as a measure of voting units, the internal vote accounting of the contract may be
compromised if voting units are transferred in any external flow by following a different formula. For example, when
implementing the hook for ERC20, the number of voting units transferred should match the formula given by the
`get_voting_units` implementation.

```

Reviewer: This warning should be helpful to users.

==== Internal functions

[.contract-item]
[[VotesComponent-_delegate]]
==== `[.contract-item-name]#++_delegate++#++(ref self: ComponentState<TContractState>, account: ContractAddress, delegatee: ContractAddress)++` [.item-kind]#internal#

Delegates all of ``account``'s voting units to `delegatee`.

Emits a {DelegateChanged} event.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-move_delegate_votes]]
==== `[.contract-item-name]#++move_delegate_votes++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Moves delegated votes from one delegate to another.

May emit one or two {DelegateVotesChanged} events.

[.contract-item]
[[VotesComponent-transfer_voting_units]]
==== `[.contract-item-name]#++transfer_voting_units++#++(ref self: ComponentState<TContractState>, from: ContractAddress, to: ContractAddress, amount: u256)++` [.item-kind]#internal#

Transfers, mints, or burns voting units.

To register a mint, `from` should be zero. To register a burn, `to`
should be zero. Total supply of voting units will be adjusted with mints and burns.

> Special care must be taken to ensure that this function is called correctly. In `ERC20` and `ERC721` this is done usually via the hooks.

Reviewer: ```suggestion
CAUTION: If voting units are a function of an underlying transferrable asset (like a token), this function should be
called every time the underlying asset is transferred to keep the internal accounting of voting power in sync. For
ERC20 and ERC721 tokens, this is usually achieved by using hooks.
```

I think this is a bit more informative. WDYT?

Author: Changed it slightly and used WARNING to be more consistent with the rest of the component, but yes agreed
